[HttpKernel] Fix session handling: decouple "save" from setting response "private"

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Fixes https://github.com/symfony/symfony/pull/25583#issuecomment-355717344 from @Tobion, and provides extra laziness for the "session" service, related to https://github.com/symfony/recipes/pull/333.

(deps=high failure will be fixed by merging to upper branches.)

nicolas-grekas committed:
[HttpKernel] Fix session handling: decouple "save" from setting response "private"

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Request.php:
```diff
@@ -863,6 +868,14 @@ public function setSession(SessionInterface $session)
+    /**
+     * @internal
+     */
+    public function setSessionFactory(callable $factory)
```
This provides extra laziness, required by https://github.com/symfony/recipes/pull/333

Pierstoval commented on src/Symfony/Component/HttpKernel/EventListener/AbstractSessionListener.php:
```diff
@@ -30,12 +30,9 @@ public function onKernelRequest(GetResponseEvent $event)
-        if (null === $session || $request->hasSession()) {
-            return;
+        if (!$request->hasSession()) {
+            $request->setSessionFactory(function () { return $this->getSession(); });
```
Wouldn't it be more secure to isolate the listener in the closure by injecting it via a `use` instead of using `$this` that can be changed before executing the closure? Maybe I'm being paranoid, so just asking 😄

nicolas-grekas replied:
Wow we never closed that far. Not sure it makes sense.
Here particularly there is nothing to close: the closure cannot be accessed.

keradus commented on src/Symfony/Component/HttpKernel/EventListener/SaveSessionListener.php:
```diff
@@ -44,19 +45,22 @@
  */
 class SaveSessionListener implements EventSubscriberInterface
 {
+    private $container;
```
please add explicit typehint for this property, typehint for constructor parameter is not the same as typehint for property

nicolas-grekas replied:
We actually removed such duplicate comments from the code base.

keradus replied:
:-1: on that, when one would do sth later with this code, he need to remember that perhaps he need to add that typehinting (eg, removing obtaining the parameter via constructor, changing visibility to public and so). I agree that for private it is no big deal, yet still.
And we do had already issues with field that was assumed to be type X, but in the end it was `null|X`, just not documented...

nicolas-grekas replied:
I respect your opinion, but this is not the proper place to discuss this. Right now, this is just consistent with the code base.

keradus replied:
I treat it same as with params, as long as we cannot typehint it on lang level, we do it in docblock ;)
Would be great if we would be able to do it with any var, like here with the property, but we are not there yet ;)
Well, I'm bet I just saw too many fuckups so I'm extra defensive...

nicolas-grekas commented:
I'm going to split the extra-laziness part for 4.1

Status: needs work

nicolas-grekas commented:
Extra-laziness removed, PR is now strictly a bug fix.

Status: needs review

sroze reviewed (commented):
Wouldn't calling the `migrate` and `save` methods on the `Session` object also start the session? i.e. we would need to set `hasStarted` to `true` ?

It's also don't really get the difference between `isStarted` and `hasStarted` now 🤔Isn't it having some logic of the storage down to the Session object?

dmaicher commented on src/Symfony/Component/HttpKernel/Tests/EventListener/SaveSessionListenerTest.php:
```diff
@@ -45,9 +45,5 @@ public function testSessionSavedAndResponsePrivate()
         $request->setSession($session);
```
The test method can now be named `testSessionSaved()` instead of `testSessionSavedAndResponsePrivate()`?

nicolas-grekas replied:
fixed

nicolas-grekas commented:
> Wouldn't calling the migrate and save methods on the Session object also start the session

fixed for "migrate", added "invalidate" also - "save" doesn't start the session.

> It's also don't really get the difference between isStarted and hasStarted

isStarted: is the session started NOW?
hasStarted: has the session be started now OR BEFORE?

nicolas-grekas commented:
Implementation updated to not require any change to the `Session` class. On the other side, the behavior is now specific to implementations of `Session` and `NativeSessionStorage`.

nicolas-grekas commented:
I finally figured out a decent way to implement this, here we are.

sroze commented on src/Symfony/Component/HttpFoundation/Session/Session.php:
```diff
@@ -29,6 +29,7 @@ class Session implements SessionInterface, \IteratorAggregate, \Countable
     private $flashName;
     private $attributeName;
     private $data = array();
+    private $wasStarted;
```
Can we rename the property and the method to `hasBeenStarted` instead of `wasStarted`? It feels to me that `was` implies it isn't "started" anymore.

nicolas-grekas replied:
renamed

sroze reviewed (commented):
Yup, much better! I still wonder something... you use this `wasStarted` argument on the `SessionBagProxy` as a sort of hidden observable. Shouldn't we actually be able to observe when the bag is initialized, and that's it? I'd properly decouple your logic (ATM `wasStarted` in the `SessionBagProxy` makes not much sense if you just look at it)

nicolas-grekas commented:
> wasStarted argument on the SessionBagProxy as a sort of hidden observable

that's exactly what the `SessionBagProxy` is for (and why it's internal also)

> I'd properly decouple your logic

feel free to submit an alternate PR (but keep in mind this should remain a bug fix, no time for refactorizations.)

sroze commented:
> that's exactly what the SessionBagProxy is for (and why it's internal also)

That's probably a good reason not to refactor 👍

nicolas-grekas commented:
and now green

Tobion commented on src/Symfony/Component/HttpKernel/EventListener/AbstractSessionListener.php:
```diff
@@ -38,10 +40,30 @@ public function onKernelRequest(GetResponseEvent $event)
+            return;
+        }
+
+        if ($session->isStarted() || ($session instanceof Session && ((\method_exists(Session::class, 'hasBeenStarted') && $session->hasBeenStarted()) || (\method_exists(Session::class, 'isEmpty') && !$session->isEmpty())))) {
```
Why does it need to maek the response uncachable when the session has been started but there is no data?

nicolas-grekas replied:
Because this is the semantic of the cache control header in HTTP: if the session is used, whatever it contains, then it means the page can be different for another user, and thus the page cannot be cached.
That's not new at all.

nicolas-grekas replied:
Hum, your question might be related to the isEmpty call on this line.
In fact this call is only required for maximum compatibility with soon-legacy versions of HttpKernel, where hasBeenStarted is not available.
We could also bump the minimum version requirement from 3.3 to 3.4.4.
I'll do that :)

nicolas-grekas replied:
Hum, your question might be related to the isEmpty call on this line.
In fact this call is only required for maximum compatibility with soon-legacy versions of HttpKernel, where hasBeenStarted is not available.
We could also bump the minimum version requirement from 3.3 to 3.4.4.
I'll do that :)

nicolas-grekas commented:
Minimum version of http-foundation bumped.

fabpot commented:
Thank you @nicolas-grekas.
